### PR TITLE
fix: identify dependencies that might require the --no-tree-shake-icons flag

### LINKED ISSUE #6197: Not able to add custom paramters to flutter build in firebase hosting
### [REQUIRED] Environment info

**firebase-tools:** 12.4.6

**Platform:** Windows

### [REQUIRED] Steps to reproduce

1. Create a Flutter project that supports web and initialize it with Firebase & firebase hosting
2. The app should contain dynamic icons, so that when normally running `flutter build web`, you'd have to add the `--no-tree-shake-icons` parameter for it to work
3. Run `firebase deploy`
### [REQUIRED] Expected behavior
It should deploy the app to firebase and recognize, that it has to add this flag OR an option to add custom flags for the build from the `firebase deploy` command, e.g. `firebase deploy --custom "--no-tree-shake-icons"`. The second option would fix the actual behavior for other cases.

### [REQUIRED] Actual behavior
Getting an error that this application cannot tree shake icon.

Logs:
```

This application cannot tree shake icons fonts. It has non-constant instances of IconData at the following locations:
<List of file paths>
Target web_release_bundle failed: Exception: Avoid non-constant invocations of IconData or try to build again with --no-tree-shake-icons.
#0      throwToolExit (package:flutter_tools/src/base/common.dart:10:3)
#1      IconTreeShaker._findConstants (package:flutter_tools/src/build_system/targets/icon_tree_shaker.dart:307:7)
<asynchronous suspension>
#2      IconTreeShaker._getIconData (package:flutter_tools/src/build_system/targets/icon_tree_shaker.dart:108:45)
<asynchronous suspension>
#3      IconTreeShaker.subsetFont (package:flutter_tools/src/build_system/targets/icon_tree_shaker.dart:172:5)
<asynchronous suspension>
#4      copyAssets.<anonymous closure> (package:flutter_tools/src/build_system/targets/assets.dart:127:25)
<asynchronous suspension>
#5      Future.wait.<anonymous closure> (dart:async/future.dart:525:21)
<asynchronous suspension>
#6      copyAssets (package:flutter_tools/src/build_system/targets/assets.dart:105:3)
<asynchronous suspension>
#7      WebReleaseBundle.build (package:flutter_tools/src/build_system/targets/web.dart:385:29)
<asynchronous suspension>
#8      _BuildInstance._invokeInternal (package:flutter_tools/src/build_system/build_system.dart:847:9)
<asynchronous suspension>
#9      Future.wait.<anonymous closure> (dart:async/future.dart:525:21)
<asynchronous suspension>
#10     _BuildInstance.invokeTarget (package:flutter_tools/src/build_system/build_system.dart:785:32)
<asynchronous suspension>
#11     FlutterBuildSystem.build (package:flutter_tools/src/build_system/build_system.dart:614:16)
<asynchronous suspension>
#12     WebBuilder.buildWeb (package:flutter_tools/src/web/compile.dart:81:34)
<asynchronous suspension>
#13     BuildWebCommand.runCommand (package:flutter_tools/src/commands/build_web.dart:191:5)
<asynchronous suspension>
#14     FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:1336:27)
<asynchronous suspension>
#15     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
<asynchronous suspension>
#16     CommandRunner.runCommand (package:args/command_runner.dart:212:13)
<asynchronous suspension>
#17     FlutterCommandRunner.runCommand.<anonymous closure> (package:flutter_tools/src/runner/flutter_command_runner.dart:296:9)
<asynchronous suspension>
#18     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
<asynchronous suspension>
#19     FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:244:5)
<asynchronous suspension>
#20     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:91:9)
<asynchronous suspension>
#21     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
<asynchronous suspension>
#22     main (package:flutter_tools/executable.dart:91:3)
<asynchronous suspension>

Compiling lib\main.dart for the Web...                           2.700ms
Exception: Failed to compile application for the Web.

Error: Unable to build your Flutter app
```

### PR DESCRIPTION
**Pull request by emwp**
### Description

Related to this issue: https://github.com/firebase/firebase-tools/issues/6197
Trying to identify dependencies that might require the --no-tree-shake-icons build flag.

Unfortunately can not allow users to pass arbitrary/custom flags when running the `deploy` function, so tried to identify dependencies that are likely to require the flag to build.

### Scenarios Tested
- Building with dependencies that would add the flag to the build command.
- Building with no dependencies that would add the flag to the build command.

fixes #6197

### CODE REVIEW

File: src/frameworks/flutter/index.ts [hunk 1]
```diff
@@ -50,9 +50,13 @@ export function init(setup: any, config: any) {
   return Promise.resolve();
 }
 
-export function build(cwd: string): Promise<BuildResult> {
+export async function build(cwd: string): Promise<BuildResult | undefined> {
```
Comment (leoortizz) [inline src/frameworks/flutter/index.ts:53]: the return type didn't change

```suggestion
export async function build(cwd: string): Promise<BuildResult> {
```

File: src/frameworks/flutter/utils.ts [hunk 2]
```diff
@@ -8,3 +10,27 @@ export function assertFlutterCliExists() {
       "Flutter CLI not found, follow the instructions here https://docs.flutter.dev/get-started/install before trying again.",
     );
 }
+
+export async function getTreeShakeFlag(): Promise<string> {
+  const pubSpecPath = "./pubspec.yaml";
+  let pubSpec: Record<string, any> = {};
+  try {
+    const pubSpecBuffer = await readFile(pubSpecPath);
```
Comment (leoortizz) [inline src/frameworks/flutter/utils.ts:18]: we should use a path relative to the project root as the project can be located in a subdir, as we do in discover https://github.com/firebase/firebase-tools/blob/cc6744f3e2bbd4fae0d08b1bf4ef999ce5c22fe0/src/frameworks/flutter/index.ts#L16-L20

File: src/frameworks/flutter/utils.ts [hunk 3]
```diff
@@ -8,3 +10,27 @@ export function assertFlutterCliExists() {
       "Flutter CLI not found, follow the instructions here https://docs.flutter.dev/get-started/install before trying again.",
     );
 }
+
+export async function getTreeShakeFlag(): Promise<string> {
```
Comment (leoortizz) [inline src/frameworks/flutter/utils.ts:14]: I think we can make it generic like `getAdditionalBuildArgs` and make it return an array of strings, then we can accommodate more use cases in the future if needed

**Review by leoortizz — CHANGES_REQUESTED** [review]

File: src/frameworks/flutter/utils.ts [hunk 4]
```diff
@@ -8,3 +10,27 @@ export function assertFlutterCliExists() {
...
+  }
+
+  const treeShakePackages = [
+    "material_icons_named",
+    "material_symbols_icons",
+    "material_design_icons_flutter",
+    "flutter_iconpicker",
+    "font_awesome_flutter",
+    "ionicons_named",
+  ];
```
Comment (leoortizz) [inline src/frameworks/flutter/utils.ts:32]: can you document why we need this? Probably linking to the issue #6197 is good enough

**Review by jamesdaniels — APPROVED** [review]
Thanks. LGTM. Let's add changelog, i'll let @leoortizz shepherd the final approval and merge

File: src/frameworks/flutter/index.ts [hunk 5]
```diff
@@ -1,23 +1,18 @@
...
 
 export const name = "Flutter Web";
 export const type = FrameworkType.Framework;
 export const support = SupportLevel.Experimental;
 
 export async function discover(dir: string): Promise<Discovery | undefined> {
-  if (!(await pathExists(join(dir, "pubspec.yaml")))) return;
-  if (!(await pathExists(join(dir, "web")))) return;
-  const pubSpecBuffer = await readFile(join(dir, "pubspec.yaml"));
-  const pubSpec = yaml.parse(pubSpecBuffer.toString());
```
Comment (leoortizz) [inline src/frameworks/flutter/index.ts:20]: we still need to return undefined if pubspec or web dir doesn't exist here

**Review by leoortizz — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 6]
```diff
@@ -1 +1,2 @@
 - Added App Hosting as an option for firebase init. (#7803)
+- Parsing pubspec.yaml to detect dependencies that might require the --no-tree-shake-icons flag in order to build. (#7724)
```
Comment (leoortizz) [inline CHANGELOG.md:2]: nit: I think we should mention Flutter web in the changelog
```suggestion
- Fixed Flutter web apps that might require --no-tree-shake-icons flag in order to build. (#7724)
```

**Review by leoortizz — APPROVED** [review]
LGTM!

File: CHANGELOG.md [hunk 7]
```diff
@@ -0,0 +1,2 @@
+- Added App Hosting as an option for firebase init. (#7803)
```
Comment (leoortizz) [inline CHANGELOG.md:1]: this is part of the previous release https://github.com/firebase/firebase-tools/releases/tag/v13.22.0
```suggestion
```

**Review by leoortizz — CHANGES_REQUESTED** [review]

**Review by leoortizz — APPROVED** [review]

**Pull request merged**